source/slang/slang-check-decl.cpp

```diff
@@ -1443,6 +1443,325 @@ namespace Slang
+        //
+        List<Expr*> synArgs;
+        for( auto paramDeclRef : getParameters(requiredMemberDeclRef) )
+        {
+            auto paramType = getType(m_astBuilder, paramDeclRef);
+
+            // For each parameter of the requirement, we create a matching
+            // parameter (same name and type) for the synthesized method.
+            //
+            auto synParamDecl = m_astBuilder->create<ParamDecl>();
+            synParamDecl->nameAndLoc = paramDeclRef.getDecl()->nameAndLoc;
+            synParamDecl->type.type = resultType;
+
+            // We need to add the parameter as a child declaration of
+            // the method we are building.
+            //
+            synParamDecl->parentDecl = synFuncDecl;
+            synFuncDecl->members.add(synParamDecl);
+
+            // For each paramter, we will create an argument expression
+            // for the call in the function body.
+            //
+            auto synArg = m_astBuilder->create<VarExpr>();
+            synArg->declRef = makeDeclRef(synParamDecl);
+            synArg->type = paramType;
+            synArgs.add(synArg);
+        }
+
+        // Required interface methods can be `static` or non-`static`,
+        // and non-`static` methods can be `[mutating]` or non-`[mutating]`.
+        // All of these details affect how we introduce our `this` parameter,
+        // if any.
+        //
+        ThisExpr* synThis = nullptr;
+        if( !requiredMemberDeclRef.getDecl()->hasModifier<HLSLStaticModifier>() )
+        {
+            // For a non-`static` requirement, we need a `this` parameter.
+            //
+            synThis = m_astBuilder->create<ThisExpr>();
+
+            // The type of `this` in our method will be the type for
+            // which we are synthesizing a conformance.
+            //
+            synThis->type.type = context->conformingType;
+
+            if( requiredMemberDeclRef.getDecl()->hasModifier<MutatingAttribute>() )
+            {
+                // If the interface requirement is `[mutating]` then our
+                // synthesized method should be too, and also the `this`
+                // parameter should be an l-value.
+                //
+                synThis->type.isLeftValue = true;
+
+                auto synMutatingAttr = m_astBuilder->create<MutatingAttribute>();
+                synFuncDecl->modifiers.first = synMutatingAttr;
+            }
+        }
+
+        // The body of our synthesized method is going to try to
+        // make a call using the name of the method requirement (e.g.,
+        // the name `increment` in our example at the top of this function).
+        //
+        // The caller already passed in a `LookupResult` that represents
+        // an attempt to look up the given name in the type of `this`,
+        // and we really just need to wrap that result up as an overloaded
+        // expression.
+        //
+        auto synBase = m_astBuilder->create<OverloadedExpr>();
```

Comment:
Having to emit checked AST directly is something that makes complex translation logic difficult to implement. I wonder if we can build some infra to make it easy, e.g.

`newAst = compileAST("return %1(%2);", funcAst, argsAst);`

Given our existing architecture, this kind of flow may not be possible...

Reply:
I agree 100% that manually building checked AST is a mess. Your approach of hijacking the parser to get it done could work, but probably requires a lot of infrastructure we don't have today.

One thing that I think can/does help is to frame our semantic checking in terms of "abstract interpretation," with a split between:

* Abstract interpretation functions that take checked syntax as input and produce checked syntax as output, potentially emitting diagnostics along the way (and/or producing an `ErrorExpr` or similar)

* AST visitor functions which orchestrate the post-order traversal of first checking sub-terms and then invoking an appropriate abstract interpretation function to do the main work and produce output syntax.

In the current semantic checking you can see some rudiments of this approach with stuff like `coerce()`, which handles implicit type conversions on already-checked syntax, or `ResolveInvoke()` which handles overload resolution for a call where all the sub-expressions are already checked.

If we wanted to go fully down that kind of road we might find that we need/want a completely separate AST hierarchy for checked vs. unchecked syntax, so that semantic checking produces a new AST rather than mutating an existing one in-place.

Certainly this one use case isn't quite enough to motivate any massive architecture changes just yet.